### PR TITLE
Node tool's editor panel UX fixes

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This PR fixes and improves the UX of the QGIS 3 revamped node tool specifically around its node editor panel. The goal is to make the node editor panel more discoverable, and eases its use and access.

The main fixes/improvements are:
- when the node editor tool is activated, the node editor panel shows up
- when the node editor panel isn't bound to a feature, show an hint label to tell users how to edit a feature via the node editor panel's table
- remove the right-click menu in favor of right-click binding a given feature to the node editor panel (faster/simpler workflow)
- when a feature is bound to the node editor panel, selecting nodes by dragging a rectangle on the canvas will only select vertices from that bound feature (much, much faster than using the selection tools, and adds a nice logic to the binding of a feature to the node editor panel)
- right click on a feature that's already bound to the node editor panel unbinds it (essentially, right-click becomes a toggle when used sequentially on a given feature)
- the node editor panel will properly restore its previous state (position, floating/docked/embedded, etc.)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
A bit of a background here, those fixes were discussed with @wonder-sk over google chat as a follow up to a twitter thread in which a user mentioned not having been aware that the node editor panel was part of 3.0.

**Comment by NathanW2** [comment]
Nice.  This should also close  PR I had open a while ago about this kind of
thing.

**Comment by nirvn** [comment]
Here's a GIF showing the improved UX, from the panel opening up when activating the node tool to the refined right-click behavior and vertices selection:
peek 2019-01-22 12-42

Notice the panel restores properly as en embedded dock sharing space with my layers panel too (obviously, it'll restore to whatever the user as set things up for him/her).

File: src/app/vertextool/qgsvertexeditor.cpp [hunk 1]
```diff
@@ -301,39 +302,68 @@ QgsVertexEditor::QgsVertexEditor(
...
 
   updateEditor( layer, selectedFeature );
 }
 
 void QgsVertexEditor::updateEditor( QgsVectorLayer *layer, QgsSelectedFeature *selectedFeature )
 {
-  delete mVertexModel;
+  if ( mSelectedFeature )
+  {
+    delete mVertexModel;
```
Comment (wonder-sk) [inline src/app/vertextool/qgsvertexeditor.cpp]: I think we should also set mVertexModel to nullptr here to make sure we do not keep a dangling pointer in some case

File: src/app/vertextool/qgsvertextool.cpp [hunk 2]
```diff
@@ -1047,30 +1068,51 @@ void QgsVertexTool::onCachedGeometryDeleted( QgsFeatureId fid )
 void QgsVertexTool::showVertexEditor()  //#spellok
 {
   QgsPointLocator::Match m = mLastMouseMoveMatch;
-  if ( !m.isValid() || !m.layer() )
-    return;
-
-  mSelectedFeature.reset( new QgsSelectedFeature( m.featureId(), m.layer(), mCanvas ) );
-  for ( int i = 0; i < mSelectedVertices.length(); ++i )
+  if ( m.isValid() || m.layer() )
```
Comment (wonder-sk) [inline src/app/vertextool/qgsvertextool.cpp]: this should probably be && instead of ||  ...?

File: src/app/vertextool/qgsvertextool.cpp [hunk 3]
```diff
@@ -1047,30 +1068,51 @@ void QgsVertexTool::onCachedGeometryDeleted( QgsFeatureId fid )
...
-    mVertexEditor.reset( new QgsVertexEditor( m.layer(), mSelectedFeature.get(), mCanvas ) );
-    QgisApp::instance()->addDockWidget( Qt::LeftDockWidgetArea, mVertexEditor.get() );
+    mVertexEditor.reset( new QgsVertexEditor( m.layer() ? m.layer() : currentVectorLayer(), mSelectedFeature ? mSelectedFeature.get() : nullptr, mCanvas ) );
+    if ( !QgisApp::instance()->restoreDockWidget( mVertexEditor.get() ) )
+      QgisApp::instance()->addDockWidget( Qt::LeftDockWidgetArea, mVertexEditor.get() );
+
     connect( mVertexEditor.get(), &QgsVertexEditor::deleteSelectedRequested, this, &QgsVertexTool::deleteVertexEditorSelection );
     connect( mVertexEditor.get(), &QgsVertexEditor::editorClosed, this, &QgsVertexTool::cleanupVertexEditor );
+
+    QTimer::singleShot( 100, this, [ = ] { mVertexEditor->show(); mVertexEditor->raise(); } );
```
Comment (wonder-sk) [inline src/app/vertextool/qgsvertextool.cpp]: why is it necessary to have 100ms delay? would be good to clarify in a comment

File: src/gui/qgsvertexmarker.cpp [hunk 4]
```diff
@@ -58,7 +59,8 @@ void QgsVertexMarker::setPenWidth( int width )
 
 void QgsVertexMarker::paint( QPainter *p )
 {
-  qreal s = ( mIconSize - 1 ) / 2.0;
+  int scaledIconSize = QgsGuiUtils::scaleIconSize( mIconSize );
+  qreal s = ( scaledIconSize - 1 ) / 2.0;
```
Comment (wonder-sk) [inline src/gui/qgsvertexmarker.cpp]: I am wondering if this should not be considered as API break. In some plugins I have been expecting that the icon size is still in pixels, so the plugin code does scaling based on DPI by itself for vertex markers and rubber bands...

This change should be probably consistent with QgsRubberBand, so both use dpi "independent" pixel sizes / widths. Maybe best to defer to a separate PR?

**Review by wonder-sk — COMMENTED** [review]
Great stuff @nirvn 

I have compiled it and tested and the new behavior feels good.

Here are my thoughts about the desired final behavior (not necessarily to be done in this PR):

- with an active/locked feature:
  - do not allow moving/adding vertices on any other feature
  - right click should unlock it (?)
  - selection of vertices and selection in vertex editor should be synchronized (there should be no "blue square" marker)
- not only right-clicking "edge" to select a feature - it should be possible also to right-click inside polygon 
 to select it (?)

File: src/gui/qgsvertexmarker.cpp (hunk 4 above)
Comment (nirvn) [reply src/gui/qgsvertexmarker.cpp]: @wonder-sk , OK, I'll do a separate PR.

**Comment by nirvn** [comment]
@wonder-sk , thanks for the review, comments addressed.

Indeed, we needed to prevent vertex&edge move and vertex addition when a feature is bound to the editor panel, that makes perfect sense, and has been implemented (woupidou).

As for your other suggestions, I think those are valid UX deficiencies that are worth fixing (especially the synching across table and canvas of selected vertices). I was hoping you'd kick in for that :smile:

**Comment by nirvn** [comment]
Unrelated travis failure, merging.

**Comment by bstroebl** [comment]
@nirvn 
First: Thank you for this improvement of the vertex tool, especially the right-click selection.
Second: I did not test this commit yet (it is still not in nighly build).

> Indeed, we needed to prevent vertex&edge move and vertex addition when a feature is bound to the editor panel, that makes perfect sense, and has been implemented (woupidou).

BUT: if topological editing is enabled I would expect nodes to be added in other features, too, and common nodes to be moved together. Same must be true for nodes in other layers while the vertex tool is in `All layers` mode. To use the node-editor panel IMHO does not automatically mean I want to only edit this particular feature.

> right click on a feature that's already bound to the node editor panel unbinds it (essentially, right-click becomes a toggle when used sequentially on a given feature)

How about: right click on a feature that's already bound to the node editor panel unbinds it **and binds the next feature in row to the editor panel** (if there is one). Thus editing overlapping polygon features becomes easier. Imagine you have two overlapping features. With the current implementation (as I understand it is working) you will always bind the same feature (the one on top) and can hardly bind the one on the bottom.

**Comment by wonder-sk** [comment]
@bstroebl yesterday I went through some vertex tool tickets and seen your request to be able to cycle through overlapping features on one location and thought the same thing - repeated right clicks could cycle through different locked features - that should fit nicely into the work Mathieu has done

**Comment by bstroebl** [comment]
I tested with QGIS bba67ae65d
Locking works fine, I like the vertices to be displayed with unfilled squares again! There are still issues, I am afraid:
1. Unlocking does only work by locking another feature or right-clicking on the locked feature. I would expect to unlock by right-clicking in an area away from my features, too.
2. The selection between the vertex panel and the map canvas is not synchronised:
a) selecting a vertex in the canvas does not select the respective row in the panel
b) selecting a vertex in the canvas marks it with a blue filled circle (additional to the unfilled red square) whereas when selecting a row in the panel the respective vertex is marked with a blue unfilled square
3. Selecting a vertex with left click for moving does only work while the locked feature is flashing. If e.g. its neighbouring feature is flashing the vertex cannot be left-clicked in order to move it, although it is a shared vertex of these two features. This behaviour makes it more difficult to move vertices although it is the default behaviour when no feature is locked but IMHO when left-clicking a vertex of the locked feature it should be moveable no matter which feature is flashing because while a feature is locked the flashing indicates that I can lock the flashing feature instead of the one I am currently working on.

**Pull request merged**